### PR TITLE
DRAFT: Design doc for pgerror linter

### PR DESCRIPTION
**Pull request by bobvawter**

### CODE REVIEW

File: pkg/cmd/retlint/README.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  instruction or basic blocks.
+  * `pgErr := err.(*pgerror.Error); return err`
+  * `if _, ok := err.(*pgerror.Error); ok { return err }`
+* The value is passed-through from another `clean` function.
+  * `if _, err := knownCleanFunction(); err != nil { return err }`
+
+In the case of cyclical call-graphs, we will assume that functions
+reachable from themselves are initially `clean` and create an
+invalidation set to record the members of the cycle.  If any member of
+the set is marked as `dirty`, all functions in the set will be dirtied.
```
Comment (knz) [inline pkg/cmd/retlint/README.md]: please provide an example for this

File: pkg/cmd/retlint/README.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  declared to be returned.
+
+The linter will then perform an inductive, heuristic analysis, starting
+from the seed functions. The goal of this analysis is to classify each
+reachable function into those that always return one of the desired
+types in place of the target interface and those that do not.  In cases
+where a seed function cannot be proven to return only the desired
+concrete types, a lint error will be generated.  The lint error will
+include the seed function and at least one call-path chain to show the
+source of the undesirable type.
```
Comment (knz) [inline pkg/cmd/retlint/README.md]: you mention only two outcomes in this paragraph but the section below details 3 states. Was that intentional?

**Review by knz — COMMENTED** [review]
very cool :)

Can't wait to see this running.

**Review by madelynnblue — APPROVED** [review]
This looks super fun.

**Pull request closed without merging**